src/lib.rs

```diff
@@ -14,36 +14,43 @@ pub use no_std_net::{IpAddr, Ipv4Addr, Ipv6Addr, SocketAddr, SocketAddrV4, Socke
 /// which knows how to send AT commands to an ESP8266 WiFi module. You could have another implementation
 /// which knows how to driver the Rust Standard Library's `std::net` module. Given this trait, you can how
 /// write a portable HTTP client which can work with either implementation.
-pub trait TcpStack {
+pub trait TcpClient {
 	/// The type returned when we create a new TCP socket
 	type TcpSocket;
 	/// The type returned when we have an error
 	type Error: core::fmt::Debug;
 
-	/// Open a new TCP socket. The socket starts in the unconnected state.
-	fn open(&self) -> Result<Self::TcpSocket, Self::Error>;
+	/// Open a socket for usage as a TCP client.
+	///
+	/// The socket must be connected before it can be used.
+	///
+	/// Returns `Ok(socket)` if the socket was successfully created.
+	fn socket(&self) -> Result<Self::TcpSocket, Self::Error>;
```

Comment:
I wonder if the UdpClient trait should have a `socket` method as well for consistency?

Reply:
I think your motivation here is right - we should have a symmetric API for both UDP and TCP. 

Should we add a `socket()` method to UDP for consistency? I don't think it would be a bad idea. My only qualm is that it doesn't make much sense to have a UDP socket in an un-opened state. What does `UdpClient::socket()` do that `UdpClient::connect()` does not do?

---

An aside about why this design is necessary for the TCP client:

To explain my design here, an alternative question is as follows:
* What's the purpose in not returning a `TcpSocket` in the result of `connect()`?

I went with this approach because `connect()` is asynchronous (requires a 3-way handshake and lots of communication between client and server before connection is established), so it has to return an `nb::Result`. What happens if `connect()` needs to return `nb::WouldBlock`? In that case, it can't return the `TcpSocket`, so there's no object returned to track the state of the connection request. Should future calls to `connect()` assume state from previous calls? That seems incorrect, which is why I went with a `socket()` constructor here.

Reply:
`UdpClient:connect` configures the device with the remote address, that's all.

Reply:
> What does UdpClient::socket() do that UdpClient::connect() does not do?

socket() would allocate one of a potentially finite number of available sockets.  Right now, the trait has that action wrapped up into `bind` or `connect`

Reply:
> `UdpClient:connect` configures the device with the remote address, that's all.

Ah, I totally forgot that UDP is connection-less, so there's no data transfer and no logical "stream" between devices. It's essentially shouting into the void - that makes sense, thanks! I think it's appropriate to make a symmetric `socket()` function for UDP as well.

Reply:
I have updated all sockets to be created using socket() and then updated all other functions to accept a reference to the original socket.